worker/src/RTC/TransportCongestionControlServer.cpp

```diff
@@ -109,6 +112,162 @@ namespace RTC
+
+			// |this->packetArrivalTimes| cannot be empty since we just added one element
+			// and the last element is not deleted.
+			// RTC_DCHECK(!this->packetArrivalTimes.empty());
+			this->periodicWindowStartSeq = this->packetArrivalTimes.begin()->first;
+		}
+	}
+
+	void TransportCongestionControlServer::SendPeriodicFeedbacks()
+	{
+		// |this->periodicWindowStartSeq| is the first sequence number to include in the
+		// current feedback packet. Some older may still be in the map, in case a
+		// reordering happens and we need to retransmit them.
+		if (!this->periodicWindowStartSeq)
+		{
+			return;
+		}
+
+		for (auto begin_iterator = this->packetArrivalTimes.lower_bound(*this->periodicWindowStartSeq);
+		     begin_iterator != this->packetArrivalTimes.cend();
+		     begin_iterator = this->packetArrivalTimes.lower_bound(*this->periodicWindowStartSeq))
+		{
+			int64_t next_sequence_number = BuildFeedbackPacket(
+			  *this->periodicWindowStartSeq, begin_iterator, this->packetArrivalTimes.cend());
+
+			// If build feedback packet fail, it will not be sent.
+			if (next_sequence_number < 0)
+			{
+				// Reset and create a new feedback packet for next periodic.
+				this->transportCcFeedbackPacket.reset(new RTC::RTCP::FeedbackRtpTransportPacket(
+				  this->transportCcFeedbackSenderSsrc, this->transportCcFeedbackMediaSsrc));
+
+				return;
+			}
+
+			this->periodicWindowStartSeq = next_sequence_number;
+			SendTransportCcFeedback();
+			// Note: Don't erase items from this->packetArrivalTimes after sending, in case
+			// they need to be re-sent after a reordering. Removal will be handled
+			// by OnPacketArrival once packets are too old.
+		}
+	}
+
+	int64_t TransportCongestionControlServer::BuildFeedbackPacket(
+	  int64_t base_sequence_number,
+	  std::map<int64_t, uint64_t>::const_iterator begin_iterator,
+	  std::map<int64_t, uint64_t>::const_iterator end_iterator)
+	{
+		// Set base sequence numer and reference time(arrival time of first received packet in the feedback).
+		uint64_t ref_timestamp_ms = begin_iterator->second;
+		this->transportCcFeedbackPacket->AddPacket(
+		  static_cast<uint16_t>((base_sequence_number - 1) & 0xFFFF),
+		  ref_timestamp_ms,
+		  this->maxRtcpPacketLen);
+
+		// RTC_DCHECK(begin_iterator != end_iterator);
+		int64_t next_sequence_number = base_sequence_number;
+
+		for (auto it = begin_iterator; it != end_iterator; ++it)
+		{
+			auto result = this->transportCcFeedbackPacket->AddPacket(
+			  static_cast<uint16_t>(it->first & 0xFFFF), it->second, this->maxRtcpPacketLen);
+
+			if (RTC::RTCP::FeedbackRtpTransportPacket::AddPacketResult::SUCCESS != result)
+			{
+				// If we can't even add the first seq to the feedback packet, we won't be
+				// able to build it at all.
+				// RTC_CHECK(begin_iterator != it);
+				MS_WARN_DEV(
+				  "add fail! result:%" PRIu32 ", cur-seq:%" PRId64 ", next-seq:%" PRId64
+				  ", base-seq:%" PRId64 "",
```

Comment:
We don't write this kind of logs anywhere.

```suggestion
				  "failed [result:%" PRIu32 ", curSeq:%" PRId64 ", nextSeq:%" PRId64", baseSeq:%" PRId64 "]",
```